extension/src/background/index.ts

```diff
@@ -26,92 +26,85 @@ import {
-                    async () =>
-                        await makeLoanRequestPayload(
-                            walletName,
-                            msg.payload.collateral,
-                            msg.payload.fee_rate,
-                        ),
-                    MessageKind.LoanResponse,
-                );
-                break;
-            case MessageKind.SignAndSendSwap:
-                try {
-                    const txHex = msg.payload;
-                    const decoded = await extractTrade(walletName, txHex);
-                    swapToSign = { txHex, decoded, tabId: sender.tab!.id! };
-                    updateBadge();
-                } catch (e) {
-                    error(e);
-                    message = { kind: MessageKind.SwapTxid, direction: Direction.ToPage, error: e };
-                }
-                break;
-            case MessageKind.SignLoan:
-                try {
-                    const details = await extractLoan(walletName, msg.payload);
-                    loanToSign = { details, tabId: sender.tab!.id! };
-                    updateBadge();
-                } catch (e) {
-                    error(e);
-                    message = { kind: MessageKind.SignedLoan, direction: Direction.ToPage, error: e };
-                }
-                break;
-        }
-        return message;
-    }
+
+var swapToSign: SwapToSign | null;
+var resolveSwapSignRequest: ((txid: Txid) => void) | null;
+var rejectSwapSignRequest: ((e: any) => void) | null;
+
+var loanToSign: LoanToSign | null;
+var resolveLoanSignRequest: ((txid: Txid) => void) | null;
+var rejectLoanSignRequest: ((e: any) => void) | null;
+
+export interface RpcMessage<T extends keyof WavesProvider> {
+    type: "rpc-message";
+    method: T;
+    args: Parameters<WavesProvider[T]>;
+}
+
+/*
+ * Defines the public interface of the background script.
+ *
+ * To ensure maximum benefit from the type checker, other components like the content script should only use this function to send messages.
+ */
+export function invokeBackgroundScriptRpc(message: Omit<RpcMessage<keyof WavesProvider>, "type">): Promise<any> {
+    return browser.runtime.sendMessage({
+        type: "rpc-message",
+        ...message,
+    });
+}
+
+addRpcMessageListener("walletStatus", () => walletStatus(walletName));
+addRpcMessageListener("getBuyCreateSwapPayload", ([usdt]) => makeBuyCreateSwapPayload(walletName, usdt));
+addRpcMessageListener("getSellCreateSwapPayload", ([btc]) => makeSellCreateSwapPayload(walletName, btc));
+addRpcMessageListener("getNewAddress", () => getAddress(walletName));
+addRpcMessageListener(
+    "makeLoanRequestPayload",
+    ([collateral, feerate]) => makeLoanRequestPayload(walletName, collateral, feerate),
+);
+
+addRpcMessageListener("signAndSendSwap", ([txHex]) => {
+    extractTrade(walletName, txHex).then(decoded => {
+        swapToSign = { txHex, decoded };
+        updateBadge();
+    });
```

Comment:
Just realized that any error here will not be forwarded to the content script, need to fix!